### Midterm Project
#### Yiscah Mark
#### nyc-rolling-sales data


##### I opened the file through a google drive link and printed the top few rows so I can examine the data and make changes.

In [1]:
import gdown
import pandas as pd

# The File ID extracted from your link
file_id = '1CgY9jjeGX8ZHSazqIqRBUfHkiNVLW0yj'
url = f'https://drive.google.com/uc?id={file_id}'

# Download the file
# This saves the file as 'midterm_data.csv' in your current folder
output = 'midterm_data.csv'
gdown.download(url, output, quiet=False)

# Open the dataset
df = pd.read_csv(output)

# Verify the requirements (Min 5 columns, 50 rows)
print(f"Dataset Shape: {df.shape}")
df.head()

Downloading...
From: https://drive.google.com/uc?id=1CgY9jjeGX8ZHSazqIqRBUfHkiNVLW0yj
To: /home/1bb32383-9cd3-4863-8b90-d3cb84c7db6b/midterm_data.csv
100%|██████████| 13.6M/13.6M [00:00<00:00, 96.4MB/s]


Dataset Shape: (84548, 22)


,Unnamed: 0,BOROUGH,NEIGHBORHOOD,BUILDING CLASS CATEGORY,TAX CLASS AT PRESENT,BLOCK,LOT,EASE-MENT,BUILDING CLASS AT PRESENT,ADDRESS,...,RESIDENTIAL UNITS,COMMERCIAL UNITS,TOTAL UNITS,LAND SQUARE FEET,GROSS SQUARE FEET,YEAR BUILT,TAX CLASS AT TIME OF SALE,BUILDING CLASS AT TIME OF SALE,SALE PRICE,SALE DATE
0,4,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2A,392,6,,C2,153 AVENUE B,...,5,0,5,1633,6440,1900,2,C2,6625000,2017-07-19 00:00:00
1,5,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2,399,26,,C7,234 EAST 4TH STREET,...,28,3,31,4616,18690,1900,2,C7,-,2016-12-14 00:00:00
2,6,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2,399,39,,C7,197 EAST 3RD STREET,...,16,1,17,2212,7803,1900,2,C7,-,2016-12-09 00:00:00
3,7,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2B,402,21,,C4,154 EAST 7TH STREET,...,10,0,10,2272,6794,1913,2,C4,3936272,2016-09-23 00:00:00
4,8,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2A,404,55,,C2,301 EAST 10TH STREET,...,6,0,6,2369,4615,1900,2,C2,8000000,2016-11-17 00:00:00


##### I am printing general information about the data set including the data type of each column. There are 22 columns and 84,548 rows of data.

In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 84548 entries, 0 to 84547
Data columns (total 22 columns):
 #   Column                          Non-Null Count  Dtype 
---  ------                          --------------  ----- 
 0   Unnamed: 0                      84548 non-null  int64 
 1   BOROUGH                         84548 non-null  int64 
 2   NEIGHBORHOOD                    84548 non-null  object
 3   BUILDING CLASS CATEGORY         84548 non-null  object
 4   TAX CLASS AT PRESENT            84548 non-null  object
 5   BLOCK                           84548 non-null  int64 
 6   LOT                             84548 non-null  int64 
 7   EASE-MENT                       84548 non-null  object
 8   BUILDING CLASS AT PRESENT       84548 non-null  object
 9   ADDRESS                         84548 non-null  object
 10  APARTMENT NUMBER                84548 non-null  object
 11  ZIP CODE                        84548 non-null  int64 
 12  RESIDENTIAL UNITS               84548 non-null

##### The first thing I wanted to check for is spelling mistakes that often cause overlap of categories. By printing all unique values in each column across the set, I can see if anything is redundant due to a spelling mistake.

In [3]:
for col in df.columns:
    print(f"--- Unique values in '{col}' ---")
    print(df[col].unique())
    print("\n")

--- Unique values in 'Unnamed: 0' ---
[    4     5     6 ... 26737 26738 26739]


--- Unique values in 'BOROUGH' ---
[1 2 3 4 5]


--- Unique values in 'NEIGHBORHOOD' ---
['ALPHABET CITY' 'CHELSEA' 'CHINATOWN' 'CIVIC CENTER' 'CLINTON'
 'EAST VILLAGE' 'FASHION' 'FINANCIAL' 'FLATIRON' 'GRAMERCY'
 'GREENWICH VILLAGE-CENTRAL' 'GREENWICH VILLAGE-WEST' 'HARLEM-CENTRAL'
 'HARLEM-EAST' 'HARLEM-UPPER' 'HARLEM-WEST' 'INWOOD' 'JAVITS CENTER'
 'KIPS BAY' 'LITTLE ITALY' 'LOWER EAST SIDE' 'MANHATTAN VALLEY'
 'MIDTOWN CBD' 'MIDTOWN EAST' 'MIDTOWN WEST' 'MORNINGSIDE HEIGHTS'
 'MURRAY HILL' 'ROOSEVELT ISLAND' 'SOHO' 'SOUTHBRIDGE' 'TRIBECA'
 'UPPER EAST SIDE (59-79)' 'UPPER EAST SIDE (79-96)'
 'UPPER EAST SIDE (96-110)' 'UPPER WEST SIDE (59-79)'
 'UPPER WEST SIDE (79-96)' 'UPPER WEST SIDE (96-116)'
 'WASHINGTON HEIGHTS LOWER' 'WASHINGTON HEIGHTS UPPER' 'BATHGATE'
 'BAYCHESTER' 'BEDFORD PARK/NORWOOD' 'BELMONT' 'BRONX PARK' 'BRONXDALE'
 'CASTLE HILL/UNIONPORT' 'CITY ISLAND' 'CITY ISLAND-PELHAM STRIP'
 'CO

In [4]:
#didn't come up in the null because there is space.

#### #1 The "Ease-ment" column is empty. This was not counted originally in the count of nulls, because there is a space in each cell. With the unique values, this issue came up as the only unique value was a space. In the following cell I remove the entire column from the data set.

In [5]:
import numpy as np

# 1. Drop empty column
df.drop(columns=['EASE-MENT'], inplace=True)

#### #2 Another important change is the date column. As seen in the output of the .info function, the dates are stored as objects. The code bellow makes the dates into real values in time.

In [6]:
# Convert 'SALE DATE' to datetime objects
df['SALE DATE'] = pd.to_datetime(df['SALE DATE'])

# Verify the transformation
print(df['SALE DATE'].dtype)

datetime64[ns]


#### #3 Another change that I made is that an index column is automatically created for the data set. This data set comes already with an index column. The code below drops the index column and uses the column named "unnamed" as the index instead.

In [7]:
# Drop the column (replace 'Unnamed: 0' with the actual name if it has one)
df = df.drop(columns=['Unnamed: 0'])

# Optional: Reset the internal index so it starts at 0
df = df.reset_index(drop=True)


In [8]:
#printing the first few rows to ensure the changes worked.
df.head()

,BOROUGH,NEIGHBORHOOD,BUILDING CLASS CATEGORY,TAX CLASS AT PRESENT,BLOCK,LOT,BUILDING CLASS AT PRESENT,ADDRESS,APARTMENT NUMBER,ZIP CODE,RESIDENTIAL UNITS,COMMERCIAL UNITS,TOTAL UNITS,LAND SQUARE FEET,GROSS SQUARE FEET,YEAR BUILT,TAX CLASS AT TIME OF SALE,BUILDING CLASS AT TIME OF SALE,SALE PRICE,SALE DATE
0,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2A,392,6,C2,153 AVENUE B,,10009,5,0,5,1633,6440,1900,2,C2,6625000,2017-07-19
1,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2,399,26,C7,234 EAST 4TH STREET,,10009,28,3,31,4616,18690,1900,2,C7,-,2016-12-14
2,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2,399,39,C7,197 EAST 3RD STREET,,10009,16,1,17,2212,7803,1900,2,C7,-,2016-12-09
3,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2B,402,21,C4,154 EAST 7TH STREET,,10009,10,0,10,2272,6794,1913,2,C4,3936272,2016-09-23
4,1,ALPHABET CITY,07 RENTALS - WALKUP APARTMENTS,2A,404,55,C2,301 EAST 10TH STREET,,10009,6,0,6,2369,4615,1900,2,C2,8000000,2016-11-17


#### #4 In the unique values I noticed 0 as a value in the "zip code" column, as well as the "year built" column. It is impossible for the values to actually be 0. I changed all the 0 cells to NaN. I took the count to see how often the issue comes up.   

In [9]:
import numpy as np

# Define the columns containing 0 as a placeholder
cols_with_zero = ['ZIP CODE', 'YEAR BUILT']

# Replace 0 with NaN
df[cols_with_zero] = df[cols_with_zero].replace(0, np.nan)

# Verification: Count how many NaNs are now in these columns
print(df[cols_with_zero].isnull().sum())

ZIP CODE       982
YEAR BUILT    6970
dtype: int64


#### #5 Out of the 84,547 rows of data, 982 is only about 1% and it would distort the data very much to make up zip codes. I dropped the columns that were missing zip codes. The "year built" consists of about 8% of the data which is more significant. In order not to loose all the information from these entries I replaced the "NaN" cells with the average year of the rest of the data set. I double checked the count of rows to see if they were dropped.

In [10]:
# 1. Drop rows with missing Zip Codes
df = df.dropna(subset=['ZIP CODE'])

# 2. Fill missing Years with the Median
year_median = df['YEAR BUILT'].median()
df['YEAR BUILT'] = df['YEAR BUILT'].fillna(year_median)

# 3. Final count check
print(f"Final Rows: {df.shape[0]}")
print(f"Final Columns: {df.shape[1]}")

Final Rows: 83566
Final Columns: 20


#### Looking more closely for spelling mistakes. It is likely that any neighborhood that appears very few times is a spelling mistake and not a real neighborhood. I didn't find any mistakes.

In [11]:
# Check for inconsistencies in Neighborhood naming
print(df['NEIGHBORHOOD'].value_counts().tail(20))

NEIGHBORHOOD
NEPONSIT                     34
ARROCHAR-SHORE ACRES         33
LIVINGSTON                   33
CONCORD-FOX HILLS            28
NEW BRIGHTON-ST. GEORGE      27
EMERSON HILL                 20
JAMAICA BAY                  19
FLUSHING MEADOW PARK         16
ROSSVILLE-RICHMOND VALLEY    14
RICHMONDTOWN-LIGHTHS HILL    14
AIRPORT LA GUARDIA            8
BLOOMFIELD                    7
ROSSVILLE-PORT MOBIL          6
CO-OP CITY                    5
DONGAN HILLS-OLD TOWN         5
FRESH KILLS                   2
VAN CORTLANDT PARK            2
BRONX PARK                    1
CITY ISLAND-PELHAM STRIP      1
EAST RIVER                    1
Name: count, dtype: int64


#### #6 The 'LAND SQUARE FEET', 'GROSS SQUARE FEET', and 'SALE PRICE' columns are all numeric values but were stored as objects. I changed the numbers in the column to integers using the .to_numeric function.

In [12]:
# Convert these three to numeric, turning dashes or spaces into NaN
cols_to_fix = ['LAND SQUARE FEET', 'GROSS SQUARE FEET', 'SALE PRICE']

for col in cols_to_fix:
    df[col] = pd.to_numeric(df[col].replace(' -  ', np.nan), errors='coerce')

#### #7 Although the Tax class at the time of sale is a number, it is just a class label with a numeric name. The number doesn't have any real value and I don't want the computer to make calculations with it so I changed it into a string.

In [13]:
# Convert Tax Class from integer to string to reflect it is a category
df['TAX CLASS AT TIME OF SALE'] = df['TAX CLASS AT TIME OF SALE'].astype(str)

# Verify the change
print(df['TAX CLASS AT TIME OF SALE'].dtype)

object


In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 83566 entries, 0 to 84547
Data columns (total 20 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   BOROUGH                         83566 non-null  int64         
 1   NEIGHBORHOOD                    83566 non-null  object        
 2   BUILDING CLASS CATEGORY         83566 non-null  object        
 3   TAX CLASS AT PRESENT            83566 non-null  object        
 4   BLOCK                           83566 non-null  int64         
 5   LOT                             83566 non-null  int64         
 6   BUILDING CLASS AT PRESENT       83566 non-null  object        
 7   ADDRESS                         83566 non-null  object        
 8   APARTMENT NUMBER                83566 non-null  object        
 9   ZIP CODE                        83566 non-null  float64       
 10  RESIDENTIAL UNITS               83566 non-null  int64         
 11  COMMERC

In [15]:
# Save the cleaned dataframe to a CSV file
# index=False prevents pandas from creating a new "Unnamed: 0" column
df.to_csv('NYC_Rolling_Sales_Cleaned.csv', index=False)

print("File saved successfully.")

File saved successfully.


#### The dataset I used, "Nyc-rolling-sales data," is a public dataset that was published by the New York City Department of Finance. This is a trustworthy enough source for me to rely on in terms of integrity. I acquired it very ethically, straight from the website. Regarding the ethical considerations of my data, as always, any distortions or changes can distort the truth and increase bias. I will review the changes made to my dataset and discuss the implications. 
#### The first change I made was to remove the "Ease-Ment" column. Every row was missing, so this obviously isn't usable information, and there was no issue with deleting it. A technical change changed the date column from an object to a datetime object. This is so the date can be chronologically ordered and used as intended. Another change to the column's data type was made to improve data accuracy: the tax class number was changed from a numeric value to a string. This is important so the model doesn't make an order or levels in the class and just keeps it as a numeric label. The "Land Square feet", "Gross Square Feet", and "Sale Price" columns were all converted to integers so they can be calculated. 
#### The biggest ethical risk was handling missing values in the "Year Built" and "Zip Code" columns. In the zip code column, I removed those rows because entering new false zip codes would have further distorted the data, and they accounted for only 1% of the data. The issue now is that it is possible there is a pattern or reason the zip code was left out; maybe lower-income households have a harder time keeping accurate records, and removing all those entries makes the class very underrepresented. All empty cells in the year-built column were set to the average year of sale for the other houses in the dataset. This is good because I did not remove the affected entries, so as not to lose the rest of the information. The problem with using the average is that, if the analysis is a time-series model, the middle year would be grossly inaccurate. For a prediction model with the year as just one of the variables, it is okay to set the year to the median so the other variables can be used. Another issue is that most of the houses that are missing a year may have been sold a very long time ago. In that case, the old houses are extremely underrepresented. What I could do to go the extra mile and maintain a very high level of integrity is to document clearly that the year-average was substituted for empty values. As for the zip code, I can cross-reference the block and lot numbers with a secondary city database to find it.
####  Some Legal Guidelines include the Fair Housing Act, which means I can’t use the data for discrimination, and the Privacy Protections rules, which means I have to do more cleaning to remove any identifying information to protect the privacy of the people in the dataset. 


Documentation:
I used gemini to get snippets of code. I wrote all the markdown cells and observations by myself.